pkg/state/steps_uki.go

```diff
@@ -557,3 +560,113 @@ func (s *State) UKIMountESPPartition(g *herd.Graph, opts ...herd.OpOption) error
+				Bytes: cert.Raw,
+			}
+			publicKeyPem := pem.EncodeToMemory(&publicKeyBlock)
+			err := os.WriteFile(filepath.Join(s.path(cnst.VerityCertDir), fmt.Sprintf("DB%d.crt", i)), publicKeyPem, 0644)
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}))...)
+}
+
+// CopySysExtensionsDagStep Copies extensions from the EFI partitions to the persistent one so they can be started.
+func (s *State) CopySysExtensionsDagStep(g *herd.Graph, opts ...herd.OpOption) error {
+	return g.Add(cnst.OpUkiCopySysExtensions, append(opts, herd.WithCallback(func(_ context.Context) error {
+		if !state.EfiBootFromInstall(internalUtils.Log) {
+			internalUtils.Log.Debug().Msg("Not copying sysextensions as we think we are booting from removable media")
+			return nil
+		}
+		// make systemd-sysext timeout early otherwise a wrong sysext will block boot
+		err := os.MkdirAll(s.path("/etc/systemd/system/systemd-sysext.service.d"), 0755)
+		if err != nil {
+			internalUtils.Log.Err(err).Msg("Creating systemd-sysext.service.d")
+			return err
+		}
+		err = os.WriteFile(s.path("/etc/systemd/system/systemd-sysext.service.d/timeout.conf"), []byte("[Service]\nTimeoutStartSec=10\n[Unit]\nJobRunningTimeoutSec=5\n"), 0644)
+		if err != nil {
+			internalUtils.Log.Err(err).Msg("Writing systemd-sysext.service.d/timeout.conf")
+			return err
+		}
+		// Write also the override for for the image policy
+		err = os.WriteFile(s.path("/etc/systemd/system/systemd-sysext.service.d/image-policy.conf"), []byte("[Service]\nExecStart=systemd-sysext refresh --image-policy=\"root=verity+signed+absent:usr=verity+signed+absent\"\nExecReload=systemd-sysext refresh --image-policy=\"root=verity+signed+absent:usr=verity+signed+absent\"\n"), 0644)
+		if err != nil {
+			internalUtils.Log.Err(err).Msg("Writing systemd-sysext.service.d/image-policy.conf")
+			return err
+		}
+		// Copy the sys extensions to the rootfs
+		// return if the source or dest dir is not there
+		if _, err := os.Stat(s.path(cnst.SourceSysExtDir)); os.IsNotExist(err) {
+			internalUtils.Log.Debug().Str("dir", s.path(cnst.SourceSysExtDir)).Msg("No sysextensions found")
+			return nil
+		}
+		if _, err := os.Stat(s.path(cnst.DestSysExtDir)); os.IsNotExist(err) {
+			_ = os.MkdirAll(s.path(cnst.DestSysExtDir), 0755)
+		}
+		err = filepath.WalkDir(s.path(cnst.SourceSysExtDir), func(path string, d fs.DirEntry, err error) error {
+			if d.IsDir() {
+				return nil
+			}
+			src := filepath.Join(cnst.SourceSysExtDir, d.Name())
+			dest := filepath.Join(cnst.DestSysExtDir, d.Name())
+
+			output, err := internalUtils.CommandWithPath(fmt.Sprintf("systemd-dissect --validate --image-policy=\"root=verity+signed+absent:usr=verity+signed+absent\" %s", src))
+			if err != nil {
+				// If the file didn't pass the validation, we don't copy it
+				internalUtils.Log.Warn().Str("src", src).Msg("Validating sysextension")
+				internalUtils.Log.Debug().Err(err).Str("src", src).Str("output", output).Msg("Validating sysextension")
```

Comment:
why log twice?

Reply:
one with debug shows the full output the other one its a warning for the user :D

Reply:
![image](https://github.com/kairos-io/immucore/assets/1447686/b3be5a6c-ef87-4277-9c57-83f13a3818ec)